File: tutorials/cpp/tutorial_thermal.cpp
#include "../../src/cpp/core/wafer.hpp"
#include "../../src/cpp/modules/geometry/geometry_manager.hpp"
#include "../../src/cpp/modules/photolithography/lithography_model.hpp"
#include "../../src/cpp/modules/metallization/metallization_model.hpp"
#include "../../src/cpp/modules/packaging/packaging_model.hpp"
#include "../../src/cpp/modules/thermal/thermal_model.hpp"
#include "../../src/cpp/renderer/vulkan_renderer.hpp"
#include <memory>
#include <iostream>

int main() {
    // Step 1: Create a wafer (300 mm diameter, 775 um thick, silicon)
    auto wafer = std::make_shared<Wafer>(300.0, 775.0, "silicon");
    GeometryManager geometry_manager;
    geometry_manager.initializeGrid(wafer, 10, 10);
    std::cout << "Initialized wafer with 10x10 grid\n";

    // Step 2: Define metal pads using photolithography
    LithographyModel lithography;
    std::vector<std::vector<int>> mask = {{1, 0, 1}, {0, 1, 0}, {1, 0, 1}};
    lithography.simulateExposure(wafer, 13.5, 0.33, mask);
    std::cout << "Applied photolithography with mask\n";

    // Step 3: Deposit copper metal layer
    MetallizationModel metallization;
    metallization.simulateMetallization(wafer, 0.5, "Cu", "pvd");
    std::cout << "Deposited 0.5 um copper layer\n";

    // Step 4: Perform electrical test to get resistance
    PackagingModel packaging;
    packaging.performElectricalTest(wafer, "resistance");
    for (const auto& prop : wafer->getElectricalProperties()) {
        std::cout << prop.first << ": " << prop.second << " Ohms\n";
    }

    // Step 5: Simulate thermal distribution
    ThermalSimulationModel thermal;
    thermal.simulateThermal(wafer, 300.0, 0.001); // 300 K ambient, 1 mA
    std::cout << "Thermal simulation completed. Max temperature: "
              << wafer->getTemperatureProfile().maxCoeff() << " K\n";

    // Step 6: Visualize results
    VulkanRenderer renderer(400, 400);
    renderer.initialize();
    renderer.render(wafer);
    std::cout << "Rendering temperature profile\n";

    return 0;
}

File: tutorials/python/tutorial_thermal.py
from semiconductor_simulator import Simulator

def main():
    print("Tutorial: Thermal Simulation in Semiconductor Simulator")
    
    # Step 1: Initialize simulator and wafer
    simulator = Simulator()
    simulator.initialize_geometry(10, 10)
    print("Step 1: Initialized 10x10 wafer grid")

    # Step 2: Pattern metal pads
    mask = [[1, 0, 1], [0, 1, 0], [1, 0, 1]]
    simulator.run_exposure(13.5, 0.33, mask)
    print("Step 2: Applied photolithography mask")

    # Step 3: Deposit metal
    simulator.run_metallization(0.5, "Cu", "pvd")
    print("Step 3: Deposited 0.5 um copper layer")

    # Step 4: Measure resistance
    simulator.perform_electrical_test("resistance")
    wafer = simulator.get_wafer()
    props = wafer.get_electrical_properties()
    for name, value in props:
        print(f"Step 4: {name}: {value} Ohms")

    # Step 5: Run thermal simulation
    simulator.simulate_thermal(300.0, 0.001)  # 300 K, 1 mA
    temp_profile = wafer.get_temperature_profile()
    print(f"Step 5: Thermal simulation complete. Max temperature: {temp_profile.max():.2f} K")

    # Step 6: Visualize using GUI
    simulator.initialize_renderer(400, 400)
    simulator.render(wafer)
    print("Step 6: Visualizing temperature profile (close window to exit)")

if __name__ == "__main__":
    main()